File: sample/src/main/swift/Sources/SampleAppCore/BoolTest.swift
//
// Created by Andrew on 2/4/18.
//

import Foundation

public struct BoolTestStruct: Codable, Hashable {
    public var yes: Bool = true
    public var no: Bool = false
    public var optional: Bool? = true
    public var optionalNil: Bool? = nil
}

public protocol BoolTestParamProtocol {
    func testParam(_ param: Bool) -> Bool
}

public protocol BoolTestReturnTypeProtocol {
    func testReturnType() -> Bool
}

public protocol BoolTestOptionalParamProtocol {
    func testOptionalParam(_ param: Bool?) -> Bool
}

public protocol BoolTestOptionalReturnTypeProtocol {
    func testOptionalReturnType() -> Bool?
}

public class BoolTest {

    public static func testYes() -> Bool {
        return true
    }

    public static func testNo() -> Bool {
        return false
    }

    public static func testParam(_ param: Bool) -> Bool {
        return param == true
    }

    public static func testReturnType() -> Bool {
        return true
    }

    public static func testOptionalParam(_ param: Bool?) -> Bool {
        return param == true
    }

    public static func testOptionalReturnType() -> Bool? {
        return true
    }

    public static func testProtocolParam(_ callback: BoolTestParamProtocol) -> Bool {
        return callback.testParam(true)
    }

    public static func testProtocolReturnType(_ callback: BoolTestReturnTypeProtocol) -> Bool {
        return callback.testReturnType()
    }

    public static func testProtocolOptionalParam(_ callback: BoolTestOptionalParamProtocol) -> Bool {
        return callback.testOptionalParam(true)
    }

    public static func testProtocolOptionalReturnType(_ callback: BoolTestOptionalReturnTypeProtocol) -> Bool? {
        return callback.testOptionalReturnType()
    }

    public static func testEncode() ->  BoolTestStruct {
        return BoolTestStruct()
    }

    public static func testDecode(_ value: BoolTestStruct) -> Bool {
        return value == BoolTestStruct()
    }

}

File: sample/src/main/java/com/readdle/swiftjava/sample/BoolTest.kt
package com.readdle.swiftjava.sample

import com.readdle.codegen.anotation.*
import java.lang.annotation.Native

@SwiftValue
data class BoolTestStruct(var yes: Boolean = true,
                          var no: Boolean = false,
                          var optional: Boolean? = true,
                          var optionalNil: Boolean? = null)

@SwiftReference
class BoolTest private constructor() {

    @SwiftDelegate(protocols = ["BoolTestParamProtocol"])
    interface BoolParamProtocol {
        @SwiftCallbackFunc
        fun testParam(param: Boolean): Boolean
    }

    @SwiftDelegate(protocols = ["BoolTestReturnTypeProtocol"])
    interface BoolReturnTypeProtocol {
        @SwiftCallbackFunc
        fun testReturnType(): Boolean
    }

    @SwiftDelegate(protocols = ["BoolTestOptionalParamProtocol"])
    interface BoolOptionalParamProtocol {
        @SwiftCallbackFunc
        fun testOptionalParam(param: Boolean?): Boolean
    }

    @SwiftDelegate(protocols = ["BoolTestOptionalReturnTypeProtocol"])
    interface BoolOptionalReturnTypeProtocol {
        @SwiftCallbackFunc
        fun testOptionalReturnType(): Boolean?
    }

    @SwiftBlock("(Bool) -> Bool")
    interface BoolBlock {
        fun call(value: Boolean): Boolean
    }

    @SwiftBlock("(Bool?) -> Bool?")
    interface OptionalBoolBlock {
        fun call(value: Boolean?): Boolean?
    }

    companion object {
        @JvmStatic
        external fun testYes(): Boolean

        @JvmStatic
        external fun testNo(): Boolean

        @JvmStatic
        external fun testParam(param: Boolean): Boolean

        @JvmStatic
        external fun testReturnType(): Boolean

        @JvmStatic
        external fun testOptionalParam(param: Boolean?): Boolean

        @JvmStatic
        external fun testOptionalReturnType(): Boolean?

        @JvmStatic
        external fun testProtocolParam(callback: BoolParamProtocol): Boolean

        @JvmStatic
        external fun testProtocolReturnType(callback: BoolReturnTypeProtocol): Boolean

        @JvmStatic
        external fun testProtocolOptionalParam(callback: BoolOptionalParamProtocol): Boolean

        @JvmStatic
        external fun testProtocolOptionalReturnType(callback: BoolOptionalReturnTypeProtocol): Boolean?

        @JvmStatic
        external fun testEncode(): BoolTestStruct

        @JvmStatic
        external fun testDecode(value: BoolTestStruct): Boolean

        @JvmStatic
        external fun testBlock(@SwiftBlock block: BoolBlock): Boolean

        @JvmStatic
        external fun testOptionalBlock(@SwiftBlock block: OptionalBoolBlock): Boolean
    }

    @Native
    var nativePointer: Long = 0

    external fun release()

}